b2/src/main/java/org/jclouds/b2/filters/RequestAuthorization.java

```diff
@@ -0,0 +1,89 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.labs.b2.filters;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.net.URI;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Resource;
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.jclouds.Constants;
+import org.jclouds.domain.Credentials;
+import org.jclouds.http.HttpException;
+import org.jclouds.http.HttpRequest;
+import org.jclouds.http.HttpRequestFilter;
+import org.jclouds.http.HttpUtils;
+import org.jclouds.labs.b2.B2Api;
+import org.jclouds.labs.b2.domain.Authorization;
+import org.jclouds.location.Provider;
+import org.jclouds.logging.Logger;
+
+import com.google.common.base.Function;
+import com.google.common.base.Supplier;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+
+@Singleton
+public class RequestAuthorization implements HttpRequestFilter {
+   @Resource
+   @Named(Constants.LOGGER_SIGNATURE)
+   Logger signatureLog = Logger.NULL;
+
+   private final Supplier<Credentials> creds;
+   private final HttpUtils utils;
+   private final LoadingCache<Credentials, Authorization> cache;
+
+   @Inject
+   public RequestAuthorization(@Provider Supplier<Credentials> creds, final B2Api b2Api, HttpUtils utils) {
+      this.creds = checkNotNull(creds, "creds");
+      this.utils = checkNotNull(utils, "utils");
+      this.cache = CacheBuilder.newBuilder()
+            .expireAfterWrite(23, TimeUnit.HOURS)
+            .build(CacheLoader.from(new Function<Credentials, Authorization>() {
+               @Override
+               public Authorization apply(Credentials creds) {
+                  return b2Api.getAuthorizationApi().authorizeAccount();
+               }
+            }));
+   }
+
+   @Override
+   public HttpRequest filter(HttpRequest request) throws HttpException {
+      utils.logRequest(signatureLog, request, ">>");
+      Credentials creds = this.creds.get();
+      Authorization auth = cache.getUnchecked(creds);
+
+      // Replace fixed URL with API URL, except for unit tests tests
+      URI endpoint = request.getEndpoint();
+      if (!endpoint.getHost().equals("localhost")) {
```

Comment:
I changed this but the alternative might not be better; we need to munge the port to get MockWebServer to work correctly.